ROADMAP.md

```diff
@@ -0,0 +1,126 @@
+# Roadmap
+
+Contains subheaders about larger improvement efforts.
+
+## `rules_haskell` support
+
+Context: [`rules_haskell`](https://github.com/tweag/rules_haskell) supports
+building Haskell targets by extending Bazel, openly maintained by Tweag.
+It would be useful if `haskell-indexer` supported generating crossreferences for
+`rules_haskell` out of the box.
+
+### Goals
+
+Provide crossreference (as Kythe entries) for both `rules_haskell`-managed
+and `nixpkgs`-sourced compilations. Latter is used by `rules_haskell` for hackage
+imports, instead of hand-rolled source imports.
+
+Crossreference data should be serveable through Kythe's language server, ready to
+be consumed by emacs's lsp-mode (with modifications if needed).
+
+Bonus goal: based on local edits, either make partial index updates possible, or
+merge local updates with the static backing index at serving time.
+
+Possible future expansions:
+  * xref `proto-lens` generated Haskell protobuf code with proto sources.
+    (https://github.com/google/haskell-indexer/issues/15)
+  * xref Haskell FFI code with C code
+    (https://github.com/google/haskell-indexer/issues/18)
+
+#### Rationale
+
+In the non-Bazel world, one might use
+[haskell-ide-engine](https://github.com/haskell/haskell-ide-engine) as a
+[LSP](https://microsoft.github.io/language-server-protocol/) server over a
+stack/cabal workspace, providing editor support (such as type sig requests,
+code navigation, completion).
+
+For Bazel/rules_haskell, one can either hand-roll the stack/cabal files in
+addition to the BUILD files to run the server. But that is duplicate work, also
+there are scaling concerns on large workspaces.
+
+So one way to get LSP support for rules_haskell is to get the Kythe entries
+using haskell-indexer, and serve them through [Kythe's language
+server](https://github.com/kythe/kythe/tree/master/kythe/go/languageserver).
+
+An alternative would be using GHC's HIE files (expected to land in GHC 8.8) to
+write an [LSIF](https://github.com/Microsoft/language-server-protocol/issues/623) index directly, which could be served by `haskell-ide-engine`
+without much overhead. This approach is not pursued here, but might be revisited
+in the future, especially if LSIF demonstrates a solid handling of cross-project
+/ cross-language references (see https://gist.github.com/robinp/76f9d3d91387da5162f773895d4e1d15).
+
+### Tasks
+
+* Decide how compilations will be extracted/analysed.
+
+  * For quick productivity boost, run analysis directly, without extraction.
+    * `rules_haskell`: maybe just indicate indexing by passing some global flag
+      (if possible with Bazel).  Alternatively use `aspects`.
```

Comment:
While it isn't currently supported, we have given some thoughts to using aspects for extraction.  The current extra_action-based mechanism is pretty awkward and hard to integrate due to the fact that it doesn't integrate with the Bazel build graph.

Aspects present a few challenges, but are probably the better mechanism.  As actions from the underlying rules were recently exposed to aspects, you can use those to more directly mirror the semantics of extra actions, but for Bazel-native rules it is up to the rule author to actually expose the action command line to Skylark (notably, C++ rules do not currently do this).  Barring that the aspect can use the exposed providers to construct the necessary extraction action, but would then be dependent on the providers to include enough information to do so.

The current infeasibility of doing so for C++ has held me back from pursuing this more actively, but I know @schroederc has been thinking about implementing it for the Java rules.

Reply:
Thanks! I suspected extra_actions are not flexible, especially for non-builtin langs which use the SpawnAction format, effectively having to parse command lines instead of access to some more structured data. I'm not sure if I would need command line access of underlying rules, will get back to you about this Thank you!